src/server/webSocketServer.ts

```diff
@@ -0,0 +1,139 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { IncomingMessage } from 'http';
+import * as ws from 'ws';
+import { helper } from '../helper';
+import { InnerLogger, logError } from '../logger';
+import { ConnectionTransport, ProtocolRequest, ProtocolResponse } from '../transport';
+
+export interface WebSocketServerDelegate {
+  onClientAttached(socket: ws): void;
+  onClientRequest(socket: ws, message: ProtocolRequest): void;
+  onClientDetached(socket: ws): void;
+  onBrowserNotification(message: ProtocolResponse): void;
+  onBrowserResponse(browserId: number, source: ws, message: ProtocolResponse): void;
+}
+
+export class WebSocketServer {
+  private _transport: ConnectionTransport;
+  private _logger: InnerLogger;
+  private _server: ws.Server;
+  private _guid: string;
+  readonly wsEndpoint: string;
+  private _bindings: (Map<any, any> | Set<any>)[] = [];
+  private _lastSeqNum = 0;
+  private _delegate: WebSocketServerDelegate;
+  private _sockets = new Set<ws>();
+  private _pendingRequests = new Map<number, { message: ProtocolRequest, source: ws | null }>();
+
+  constructor(transport: ConnectionTransport, logger: InnerLogger, port: number, delegate: WebSocketServerDelegate) {
+    this._guid = helper.guid();
+    this._transport = transport;
+    this._logger = logger;
+    this._server = new ws.Server({ port });
+    this._delegate = delegate;
+    transport.onmessage = message => this._browserMessage(message);
+    transport.onclose = () => this._browserClosed();
+    this._server.on('connection', (socket: ws, req) => this._clientAttached(socket, req));
+    const address = this._server.address();
+    this.wsEndpoint = typeof address === 'string' ? `${address}/${this._guid}` : `ws://127.0.0.1:${address.port}/${this._guid}`;
+  }
+
+  addBindings(bindings: (Map<any, any>|Set<any>)[]) {
+    this._bindings.push(...bindings);
+  }
+
+  sendMessageToBrowser(message: ProtocolRequest, source: ws): number {
+    const seqNum = ++this._lastSeqNum;
+    this._pendingRequests.set(seqNum, { message, source });
+    this._transport.send({ ...message, id: seqNum });
+    return seqNum;
+  }
+
+  sendMessageToBrowserOneWay(method: string, params: any) {
+    this._transport.send({ id: ++this._lastSeqNum, method, params });
+  }
+
+  close() {
+    this._server.close();
+  }
+
+  private _browserMessage(message: ProtocolResponse) {
+    const seqNum = message.id;
+    if (typeof seqNum === 'number') {
+      const request = this._pendingRequests.get(seqNum);
+      if (!request)
+        return;
```

Comment:
assert/debugAssert?

Reply:
Can't do - this will knock the server down.

Reply:
Done